File: Problemas/fenwicktree.py
'''
Aplicações:
    - URI 1804: Precisa-se de Matemáticos em Marte
    - OBI 2017: Arranha-céu
'''
def LSOne(S):
    return S & (-S)

class FenwickTree:
    def __init__(self, N):
        self.ft = [0] * (N + 1)

    def rsq(self, b):
        soma = 0
        while b > 0:
            soma += self.ft[b]
            b -= LSOne(b)
        return soma

    def update(self, k, v):
        while k < len(self.ft):
            self.ft[k] += v
            k += LSOne(k)

File: Problemas/FenwickTree/fenwicktree.cpp
using namespace std;
#define MAX 11234
int LSOne(int N){
    return N & (-N);
}

int ft[MAX];

void update(int k, int v){
    while(k < MAX){
        ft[k] += v;
        k += LSOne(k);
    }
}

int rsq(int b){
    int soma = 0;
    while(b > 0){
        soma += ft[b];
        b -= LSOne(b);
    }
    return soma;
}

int rsq(int a, int b){
    return (a == 0 ? rsq(b) : rsq(b) - rsq(a-1));
}